### ISSUE TITLE
IO::Handle autoflush breaks perldebugger w/PERLDB_OPTS="Frame=2" at 70c943

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#12773]
Migrated from [rt.perl.org#116769](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=116769) (status was 'resolved')

Searchable as RT116769$

**Comment by p5pRT** [comment]
### From @kentfredric

#### Created by @kentfredric

This is a bug report for perl from kentfredric@&#8203;gmail\.com\,
generated with the help of perlbug 1\.39 running under perl 5\.17\.9\.

\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
After a bit of digging I found the root\.

The problem&#8203;:

&nbsp;  PERLDB\_OPTS='Frame=2 nonstop' perl \-d \-E1

result in the perl debugger dying in a ball of fire&#8203;:

Can't use an undefined value as a symbol reference at
/home/kent/perl5/perlbrew/perls/g6791e41/lib/5\.17\.6/perl5db\.pl line
5540\.
Compilation failed in require\.
 at \-e line 0\.
&nbsp;  main&#8203;::BEGIN\(\) called at \-e line 0
&nbsp;  eval \{\.\.\.\} called at \-e line 0
BEGIN failed\-\-compilation aborted\.
 at \-e line 0\.

Upon modifying the code where the failure occurred\, it appeared that
the problem arose from $LINEINFO not being defined\.

So a git bisect was done\, and this eventually lead to the point of failure

Commit&#8203;: 70c9432b92364dee2e7c675da23ab5c74634ee5f
Author&#8203;: Shlomi Fish \<shlomif@&#8203;shlomifish\.org>  2012\-09\-30 10&#8203;:42&#8203;:02
Committer&#8203;: Ricardo Signes \<rjbs@&#8203;cpan\.org>  2012\-11\-13 03&#8203;:18&#8203;:27
&nbsp;    \[perl5db\] Convert $| to \->autoflush\(1\)\.


So I changed the first incarnation back to its original form without
the filehandle

\+    my $previous = select\($OUT\);
\+    $| = 1;                                  \# for DB&#8203;::OUT
\+    select\($previous\);
\-    $OUT\->autoflush\(1\);

And behold\, the problem vanished with 'Frame=1'

I then inserted

\+    require IO&#8203;::File;

And this gives a very similar error\, except with a backtrace\,
suggesting that the cause is the autoloading
of IO&#8203;::File at this point in the code for some reason;

PERLDB\_OPTS="Frame=1 nonstop" perl \-d \-E1
Can't use an undefined value as a symbol reference at
/home/kent/perl5/perlbrew/perls/perl\-blead/lib/5\.17\.9/perl5db\.pl line
6122\.
 at /home/kent/perl5/perlbrew/perls/perl\-blead/lib/5\.17\.9/perl5db\.pl line 6122\.
&nbsp;  DB&#8203;::print\_lineinfo\('  '\, 'entering CODE\(0x1412920\)\\x\{a\}'\) called at
/home/kent/perl5/perlbrew/perls/perl\-blead/lib/5\.17\.9/perl5db\.pl line
4042
&nbsp;  DB&#8203;::\_indent\_print\_line\_info\(\-1\, 'entering CODE\(0x1412920\)\\x\{a\}'\)
called at /home/kent/perl5/perlbrew/perls/perl\-blead/lib/5\.17\.9/perl5db\.pl
line 4063
&nbsp;  DB&#8203;::\_print\_frame\_message\(\) called at
/home/kent/perl5/perlbrew/perls/perl\-blead/lib/5\.17\.9/x86\_64\-linux/IO/File\.pm
line 127
&nbsp;  eval \{\.\.\.\} called at
/home/kent/perl5/perlbrew/perls/perl\-blead/lib/5\.17\.9/x86\_64\-linux/IO/File\.pm
line 127
&nbsp;  require IO/File\.pm called at
/home/kent/perl5/perlbrew/perls/perl\-blead/lib/5\.17\.9/perl5db\.pl line
1661
&nbsp;  require perl5db\.pl called at \-e line 0
&nbsp;  main&#8203;::BEGIN\(\) called at
/home/kent/perl5/perlbrew/perls/perl\-blead/lib/5\.17\.9/x86\_64\-linux/IO/File\.pm
line 127
&nbsp;  eval \{\.\.\.\} called at
/home/kent/perl5/perlbrew/perls/perl\-blead/lib/5\.17\.9/x86\_64\-linux/IO/File\.pm
line 127
BEGIN failed\-\-compilation aborted at
/home/kent/perl5/perlbrew/perls/perl\-blead/lib/5\.17\.9/x86\_64\-linux/IO/File\.pm
line 127\.
 at /home/kent/perl5/perlbrew/perls/perl\-blead/lib/5\.17\.9/x86\_64\-linux/IO/File\.pm
line 127\.
&nbsp;  require IO/File\.pm called at
/home/kent/perl5/perlbrew/perls/perl\-blead/lib/5\.17\.9/perl5db\.pl line
1661
&nbsp;  require perl5db\.pl called at \-e line 0
&nbsp;  main&#8203;::BEGIN\(\) called at
/home/kent/perl5/perlbrew/perls/perl\-blead/lib/5\.17\.9/x86\_64\-linux/IO/File\.pm
line 127
&nbsp;  eval \{\.\.\.\} called at
/home/kent/perl5/perlbrew/perls/perl\-blead/lib/5\.17\.9/x86\_64\-linux/IO/File\.pm
line 127
Compilation failed in require at
/home/kent/perl5/perlbrew/perls/perl\-blead/lib/5\.17\.9/perl5db\.pl line
1661\.
Compilation failed in require\.
 at \-e line 0\.
&nbsp;  main&#8203;::BEGIN\(\) called at \-e line 0
&nbsp;  eval \{\.\.\.\} called at \-e line 0
BEGIN failed\-\-compilation aborted\.
 at \-e line 0\.

<details><summary>Perl Info</summary>

```

Flags:
    category=utilities
    severity=high

Site configuration information for perl 5.17.9:

Configured by kent at Fri Feb 15 17:27:26 NZDT 2013.

Summary of my perl5 (revision 5 version 17 subversion 9) configuration:
  Snapshot of: 012528a99df1906d6feaaaf05657a53374aebc46
  Platform:
    osname=linux, osvers=3.7.6-gentoo, archname=x86_64-linux
    uname='linux katipo2 3.7.6-gentoo #26 smp sat feb 9 08:42:27 nzdt
2013 x86_64 intel(r) core(tm) i5-2410m cpu @ 2.30ghz genuineintel
gnulinux '
    config_args='-de
-Dprefix=/home/kent/perl5/perlbrew/perls/perl-blead -Dusedevel
-Aeval:scriptdir=/home/kent/perl5/perlbrew/perls/perl-blead/bin'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=undef, usemultiplicity=undef
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=define, use64bitall=define, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-fno-strict-aliasing -pipe -fstack-protector
-D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2',
    cppflags='-fno-strict-aliasing -pipe -fstack-protector'
    ccversion='', gccversion='4.7.2', gccosandvers=''
    intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t',
lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -fstack-protector'
    libpth=/lib/../lib64 /usr/lib/../lib64 /lib /usr/lib /lib64 /usr/lib64
    libs=-lnsl -lgdbm -ldb -ldl -lm -lcrypt -lutil -lc -lgdbm_compat
    perllibs=-lnsl -ldl -lm -lcrypt -lutil -lc
    libc=/lib/libc-2.15.so, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.15'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -O2 -fstack-protector'

Locally applied patches:



@INC for perl 5.17.9:
    /home/kent/perl5/perlbrew/perls/perl-blead/lib/site_perl/5.17.9/x86_64-linux
    /home/kent/perl5/perlbrew/perls/perl-blead/lib/site_perl/5.17.9
    /home/kent/perl5/perlbrew/perls/perl-blead/lib/5.17.9/x86_64-linux
    /home/kent/perl5/perlbrew/perls/perl-blead/lib/5.17.9
    .


Environment for perl 5.17.9:
    HOME=/home/kent
    LANG (unset)
    LANGUAGE (unset)
    LC_CTYPE=en_NZ.UTF8
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/kent/perl5/perlbrew/bin:/home/kent/perl5/perlbrew/perls/perl-blead/bin:/usr/local/bin:/usr/bin:/bin:/opt/bin:/usr/x86_64-pc-linux-gnu/gcc-bin/4.7.2:/usr/games/bin
    PERLBREW_BASHRC_VERSION=0.46
    PERLBREW_HOME=/home/kent/.perlbrew
    PERLBREW_MANPATH=/home/kent/perl5/perlbrew/perls/perl-blead/man
    PERLBREW_PATH=/home/kent/perl5/perlbrew/bin:/home/kent/perl5/perlbrew/perls/perl-blead/bin
    PERLBREW_PERL=perl-blead
    PERLBREW_ROOT=/home/kent/perl5/perlbrew
    PERLBREW_VERSION=0.46
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @kentfredric

I should point out that there is no way of avoiding this bug\, because 
even if you 'use IO&#8203;::File'\, calling \->autoflush brings this problem up n 
matter what I try\.

ie&#8203;: I rewrote the code to be 

sub \_autoflush \{
&nbsp;    my $fh = shift;
&nbsp;    my $o = select\($fh\);
&nbsp;    $|\+\+;
&nbsp;    select\($o\);
\}

And replaced \->autoflush everywhere with calls to that function\.

No Problem&#8203;:
\-\-
use IO&#8203;::File;
sub \_autoflush \{
&nbsp;    my $fh = shift;
&nbsp;    my $o = select\($fh\);
&nbsp;    $|\+\+;
&nbsp;    select\($o\);
\}
\-\-

No Problem&#8203;:
\-\-
require IO&#8203;::File;
sub \_autoflush \{
&nbsp;    my $fh = shift;
&nbsp;    my $o = select\($fh\);
&nbsp;    $|\+\+;
&nbsp;    select\($o\);
\}
\-\-

Problem&#8203;:
\-\-

sub \_autoflush \{
&nbsp;    require IO&#8203;::File;
&nbsp;    my $fh = shift;
&nbsp;    my $o = select\($fh\);
&nbsp;    $|\+\+;
&nbsp;    select\($o\);
\}
\-\-

No Problem&#8203;:
\-\-
use IO&#8203;::File;
sub \_autoflush \{
&nbsp;    require IO&#8203;::File;
&nbsp;    my $fh = shift;
&nbsp;    my $o = select\($fh\);
&nbsp;    $|\+\+;
&nbsp;    select\($o\);
\}
\-\-

Problem&#8203;:
\-\-
use IO&#8203;::File;
sub \_autoflush \{
&nbsp;    require IO&#8203;::File;
&nbsp;    $\_\[0\]\->autoflush\(1\);
\}
\-\-

In summary\, the only way to refactor this sanely is to do what I did\, 
and have a perl5db\.pl\-specific function for this with no external 
dependencies\.

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @kentfredric

Attached is a suggested patch which should solve this problem 
indefinitely\, while keeping the code that invokes this logic simple\.

**Comment by p5pRT** [comment]
### From @kentfredric

<details><summary>0001-lib-perl5db.pl-refactor-all-calls-to-IO-Handle-via-a.patch</summary>

```diff
From 3e994e1b3c3d940ee87e53e2aeb772c1755e4fbd Mon Sep 17 00:00:00 2001
From: Kent Fredric <kentfredric@gmail.com>
Date: Fri, 15 Feb 2013 23:24:56 +1300
Subject: [PATCH] lib/perl5db.pl: refactor all calls to IO::Handle via
 ->autoflush to a short local function.

This hopefully resovles https://rt.perl.org/rt3/Ticket/Display.html?id=116769
and any issue similar to that which I may not have been unlucky enough
to fall upon.
---
 lib/perl5db.pl | 19 ++++++++++++++-----
 1 file changed, 14 insertions(+), 5 deletions(-)

diff --git a/lib/perl5db.pl b/lib/perl5db.pl
index 54d6622..0c51d4e 100644
--- a/lib/perl5db.pl
+++ b/lib/perl5db.pl
@@ -1472,6 +1472,15 @@ use vars qw($lineinfo $doccmd);
 
 our ($runnonstop);
 
+# Local autoflush to avoid rt#116769,
+# as calling IO::File methods causes an unresolvable loop
+# that results in debugger failure.
+sub _autoflush { 
+    my $o = select($_[0]);
+    $|++;
+    select($o);
+}
+
 if ($notty) {
     $runnonstop = 1;
     share($runnonstop);
@@ -1655,7 +1664,7 @@ and if we can.
     } ## end elsif (from if(defined $remoteport))
 
     # Unbuffer DB::OUT. We need to see responses right away.
-    $OUT->autoflush(1);
+    _autoflush($OUT);
 
     # Line info goes to debugger output unless pointed elsewhere.
     # Pointing elsewhere makes it possible for slave editors to
@@ -2324,7 +2333,7 @@ sub _DB__handle_run_command_in_pager_command {
         if $pager =~ /^\|/
         && ( "" eq $SIG{PIPE} || "DEFAULT" eq $SIG{PIPE} );
 
-        OUT->autoflush(1);
+        _autoflush(\*OUT);
         # Save current filehandle, and put it back.
         $obj->selected(scalar( select(OUT) ));
         # Don't put it back if pager was a pipe.
@@ -6757,7 +6766,7 @@ sub setterm {
             open( OUT, ">$o" ) or die "Cannot open TTY '$o' for write: $!";
             $IN  = \*IN;
             $OUT = \*OUT;
-            $OUT->autoflush(1);
+            _autoflush($OUT);
         } ## end if ($tty)
 
         # We don't have a TTY - try to find one via Term::Rendezvous.
@@ -7592,7 +7601,7 @@ sub reset_IN_OUT {
     }
 
     # Unbuffer the output filehandle.
-    $OUT->autoflush(1);
+    _autoflush($OUT);
 
     # Point LINEINFO to the same output filehandle if it was there before.
     $LINEINFO = $OUT if $switch_li;
@@ -7870,7 +7879,7 @@ sub LineInfo {
         open ($new_lineinfo_fh , $stream )
             or _db_warn("Cannot open '$stream' for write");
         $LINEINFO = $new_lineinfo_fh;
-        $LINEINFO->autoflush(1);
+        _autoflush($LINEINFO);
     }
 
     return $lineinfo;
-- 
1.7.12.4


```
</details>

**Comment by p5pRT** [comment]
### From @kentfredric

Here is a test that appears to test for this behavior

**Comment by p5pRT** [comment]
### From @kentfredric

<details><summary>0002-lib-perl5db.t-Add-test-for-bug-116769-Frame-2-crashe.patch</summary>

```diff
From 5b50ed5b76d13e1d1cad8ac8dfa622c149057098 Mon Sep 17 00:00:00 2001
From: Kent Fredric <kentfredric@gmail.com>
Date: Thu, 21 Feb 2013 22:41:48 +1300
Subject: [PATCH 2/3] lib/perl5db.t: Add test for bug #116769, Frame=2 crashes
 debugger

---
 lib/perl5db.t | 9 ++++++++-
 1 file changed, 8 insertions(+), 1 deletion(-)

diff --git a/lib/perl5db.t b/lib/perl5db.t
index 174554f..9bd2453 100644
--- a/lib/perl5db.t
+++ b/lib/perl5db.t
@@ -28,7 +28,7 @@ BEGIN {
     }
 }
 
-plan(109);
+plan(111);
 
 my $rc_filename = '.perldb';
 
@@ -93,6 +93,13 @@ EOF
     my $output = runperl(switches => [ '-d' ], stderr => 1, progfile => '../lib/perl5db/t/rt-66110');
     like($output, "All tests successful.", "[perl #66110]");
 }
+# [ perl #116769] Frame=2
+{
+    local $ENV{PERLDB_OPTS} = "frame=2 nonstop";
+    my $output = runperl( switches => [ '-d' ], prog => 'print q{success}' );
+    is( $?, 0, '[perl #116769] frame=2 does not crash debugger, exit == 0' );
+    like( $output, 'success' , '[perl #116769] code is run' );
+}
 
 {
     rc(<<'EOF');
-- 
1.7.12.4


```
</details>

**Comment by p5pRT** [comment]
### From @kentfredric

Resubmitting 0001 Patch\, but without the accidental trailing whitespace\.

**Comment by p5pRT** [comment]
### From @kentfredric

<details><summary>0001-lib-perl5db.pl-refactor-all-calls-to-IO-Handle-via-a.patch</summary>

```diff
From 96aa522afe030dd9da94fa7b268532eb89352d30 Mon Sep 17 00:00:00 2001
From: Kent Fredric <kentfredric@gmail.com>
Date: Fri, 15 Feb 2013 23:24:56 +1300
Subject: [PATCH 1/4] lib/perl5db.pl: refactor all calls to IO::Handle via
 ->autoflush to a short local function.

This hopefully resovles https://rt.perl.org/rt3/Ticket/Display.html?id=116769
and any issue similar to that which I may not have been unlucky enough
to fall upon.
---
 lib/perl5db.pl | 19 ++++++++++++++-----
 1 file changed, 14 insertions(+), 5 deletions(-)

diff --git a/lib/perl5db.pl b/lib/perl5db.pl
index 54d6622..e6ee7c0 100644
--- a/lib/perl5db.pl
+++ b/lib/perl5db.pl
@@ -1472,6 +1472,15 @@ use vars qw($lineinfo $doccmd);
 
 our ($runnonstop);
 
+# Local autoflush to avoid rt#116769,
+# as calling IO::File methods causes an unresolvable loop
+# that results in debugger failure.
+sub _autoflush {
+    my $o = select($_[0]);
+    $|++;
+    select($o);
+}
+
 if ($notty) {
     $runnonstop = 1;
     share($runnonstop);
@@ -1655,7 +1664,7 @@ and if we can.
     } ## end elsif (from if(defined $remoteport))
 
     # Unbuffer DB::OUT. We need to see responses right away.
-    $OUT->autoflush(1);
+    _autoflush($OUT);
 
     # Line info goes to debugger output unless pointed elsewhere.
     # Pointing elsewhere makes it possible for slave editors to
@@ -2324,7 +2333,7 @@ sub _DB__handle_run_command_in_pager_command {
         if $pager =~ /^\|/
         && ( "" eq $SIG{PIPE} || "DEFAULT" eq $SIG{PIPE} );
 
-        OUT->autoflush(1);
+        _autoflush(\*OUT);
         # Save current filehandle, and put it back.
         $obj->selected(scalar( select(OUT) ));
         # Don't put it back if pager was a pipe.
@@ -6757,7 +6766,7 @@ sub setterm {
             open( OUT, ">$o" ) or die "Cannot open TTY '$o' for write: $!";
             $IN  = \*IN;
             $OUT = \*OUT;
-            $OUT->autoflush(1);
+            _autoflush($OUT);
         } ## end if ($tty)
 
         # We don't have a TTY - try to find one via Term::Rendezvous.
@@ -7592,7 +7601,7 @@ sub reset_IN_OUT {
     }
 
     # Unbuffer the output filehandle.
-    $OUT->autoflush(1);
+    _autoflush($OUT);
 
     # Point LINEINFO to the same output filehandle if it was there before.
     $LINEINFO = $OUT if $switch_li;
@@ -7870,7 +7879,7 @@ sub LineInfo {
         open ($new_lineinfo_fh , $stream )
             or _db_warn("Cannot open '$stream' for write");
         $LINEINFO = $new_lineinfo_fh;
-        $LINEINFO->autoflush(1);
+        _autoflush($LINEINFO);
     }
 
     return $lineinfo;
-- 
1.8.1.5


```
</details>

**Comment by p5pRT** [comment]
### From @rjbs

On Wed Mar 13 09&#8203;:20&#8203;:38 2013\, kentfredric wrote&#8203;:

Right now\, it all seems to pass\. I pushed it to smoke\-me/rjbs/debug\-kentnl\.

\-\- 
rjbs

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

On Wed Mar 13 09&#8203;:20&#8203;:38 2013\, kentfredric wrote&#8203;:

Right now\, it all seems to pass\. I pushed it to smoke\-me/rjbs/debug\-kentnl\.

\-\- 
rjbs

**Comment by p5pRT** [comment]
### From @rjbs

Thanks\, applied in blead\.

\-\- 
rjbs

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

Thanks\, applied in blead\.

\-\- 
rjbs

**Comment by p5pRT** [comment]
@rjbs - Status changed from 'open' to 'resolved'
